#Question 1.What are the five key concepts of Object-Oriented Programming (OOP)?


#Answer :

The five key concepts of Object-Oriented Programming (OOP) are:


**Encapsulation:** This concept involves wrapping both data (attributes) and the functions (methods) that interact with the data into a single unit, typically known as a class. Encapsulation also restricts access to specific details of an object, ensuring that the internal workings are hidden, protecting it from external modification or misuse.

**Abstraction:** Abstraction focuses on representing essential features of an object while omitting unnecessary details. This simplifies the design, allowing developers to work with complex systems at a higher level without needing to worry about the underlying intricacies.

**Inheritance:**  Inheritance allows one class (called a subclass) to derive properties and behavior from another class (the superclass). This facilitates code reuse and establishes a hierarchical structure between classes, making it easier to extend and maintain the code.

**Polymorphism:** Polymorphism enables objects to be treated as instances of their parent class, even when they belong to different derived classes. It can take the form of method overriding, where a subclass modifies a method from its superclass, or method overloading, where methods share the same name but differ in parameters.

**Composition:** While not always highlighted as a core concept, composition involves constructing more complex objects by combining simpler ones. This results in a "has-a" relationship between objects, allowing for greater flexibility and dynamic interactions within a system.

These principles collectively help in organizing code in a more modular, maintainable, and scalable way, improving overall software development efficiency.


#2. Write a Python class for a `Car` with attributes for `make`, `model`, and `year`. Include a method to display the car's information

#Answer:

A basic Python class for an car that has the following attributes: maker, model, and year; it also has a function to show the information about the car:

In [ ]:
class Car:
    def __init__(self, make, model, year):
        self.make = make
        self.model = model
        self.year = year

    def show_details(self):
        print(f"Details: {self.year} {self.make} {self.model}")

# Example usage:
my_vehicle = Car("BMW", "Q7", 2024)
my_vehicle.show_details()


Details: 2024 BMW Q7


The output for this example would be:
Details: 2024 BMW Q7

#3. Explain the difference between instance methods and class methods. Provide an example of each.

#Answer:

In Python, instance methods and class methods are types of methods associated with a class, but they differ in terms of what they operate on and how they are called.

**1. Instance Methods**

What they are:
These methods are tied to an individual instance of a class. They can access and modify data specific to that instance (i.e., instance attributes) and are invoked on the object itself.

How they're defined:
They always take self as the first parameter, which represents the instance calling the method.

When to use: When you need to work with or change data that belongs to a specific object (instance) of the class.


Example of an Instance Method:

In [ ]:
class Car:
    def __init__(self, make, model):
        self.make = make  # Instance-specific attribute
        self.model = model  # Instance-specific attribute

    def get_info(self):  # Instance method
        return f"This car is a {self.make} {self.model}"

# Creating an instance of the Car class
my_car = Car("BMW", "Q7")
print(my_car.get_info())  # Outputs: This car is a BMW Q7


This car is a BMW Q7


Here, get_info() is an instance method because it uses self to refer to the particular instance’s attributes (like make and model).


**2. Class Methods**


What they are:

These methods are linked to the class rather than any specific instance of the class. They work with class-level data and can modify class attributes, which are shared by all instances.

How they're defined:

They take cls as the first parameter, which refers to the class itself.

When to use:

Use them when you need to operate on or access class-level attributes.

Decorator:

Class methods are marked with the @classmethod decorator.


Example of a Class Method:

In [ ]:
class Car:
    num_wheels = 4  # Class-level attribute

    @classmethod
    def get_wheel_count(cls):  # Class method
        return f"All cars have {cls.num_wheels} wheels"

# Calling the class method
print(Car.get_wheel_count())  # Outputs: All cars have 4 wheels

All cars have 4 wheels


In this example, get_wheel_count() is a class method because it uses cls to access the class-level attribute num_wheels and does not require an instance to be called.



**Main Differences:**

Instance methods: Operate on data specific to an object and require an instance of the class.


Class methods: Operate on class-level data and don’t need an instance to be invoked. Instead, they use the class itself (cls).


This distinction helps determine whether methods should deal with individual object data or broader class-level data.


#Question 4. How does Python implement method overloading? Give an example.

#Answer:

In Python, method overloading (as seen in languages like Java or C++) is not directly supported. Python does not allow defining multiple methods with the same name but different parameter signatures. However, you can mimic this behavior by using techniques such as variable-length arguments and then checking the arguments inside the method.


Python may simulate method overloading by verifying the quantity and type of arguments given. Here's an example:

In [ ]:
class Calculator:
    def add(self, *numbers):
        if len(numbers) == 2:
            return numbers[0] + numbers[1]
        elif len(numbers) == 3:
            return numbers[0] + numbers[1] + numbers[2]
        else:
            return "Unsupported number of arguments"

# Example usage
calc = Calculator()

print(calc.add(12, 24))           # Output: 36
print(calc.add(15, 12, 18))        # Output: 45
print(calc.add(20))               # Output: Unsupported number of arguments


36
45
Unsupported number of arguments


**Explanation:**

The *numbers argument allows the method to accept any number of arguments.

The method determines the number of provided arguments and then applies logic based on that, enabling it to simulate method overloading.

This approach allows the method to handle different numbers of arguments while maintaining a single method name.

#Question 5. What are the three types of access modifiers in Python? How are they denoted?

#Answer:

In Python, access modifiers are used to control the visibility of attributes and methods in classes. There are three main types, each with its own denotation:

**1.Public:**

Denotation:
No leading underscores.

Description:

Public attributes and methods can be accessed from anywhere, both inside and outside the class. By default, all class members are public unless specified otherwise.



In [ ]:
class MyClass:
    def __init__(self):
        self.public_attribute = "Accessible everywhere"

    def public_method(self):
        return "This is a public method"


**2.Protected:**

Denotation:
A single leading underscore (_).

Description:

Attributes and methods marked with a single underscore are considered protected. They are intended for use within the class and its subclasses but can still be accessed from outside the class, though this is generally discouraged.



In [ ]:
class MyClass:
    def __init__(self):
        self._protected_attribute = "Intended for internal use"

    def _protected_method(self):
        return "This is a protected method"

**3. Private:**

Denotation:
A double leading underscore (__).

Description:

Attributes and methods with double underscores are private. They are not accessible from outside the class due to name mangling, which alters the name to prevent direct access.

In [ ]:
class MyClass:
    def __init__(self):
        self.__private_attribute = "Restricted access"

    def __private_method(self):
        return "This is a private method"

**Summary:**

Public:

No underscore (accessible from anywhere).

Protected:

Single underscore (_) (accessible within the class and its subclasses).

Private:

Double underscore (__) (accessible only within the class, with name mangling applied).

#Question 6. Describe the five types of inheritance in Python. Provide a simple example of multiple inheritance.

#Answer:

Inheritance in Python allows you to create new classes based on existing ones, enabling code reuse and the establishment of a class hierarchy. Here are the five main types of inheritance:


**1. Single Inheritance**

In single inheritance, a subclass derives from only one parent class. This is the most straightforward form of inheritance.

In [ ]:
#Example:

class Parent:
    def display(self):
        print("This is the Parent class.")

class Child(Parent):
    def show(self):
        print("This is the Child class.")

# Usage
child_instance = Child()
child_instance.display()  # Method from Parent
child_instance.show()     # Method from Child

This is the Parent class.
This is the Child class.


**2. Multiple Inheritance**

Multiple inheritance allows a subclass to inherit from more than one parent class, gaining access to attributes and methods from multiple sources.

In [ ]:
#Example:

class Father:
    def skills(self):
        print("Father's skills: Driving, Carpentry")

class Mother:
    def skills(self):
        print("Mother's skills: Cooking, Sewing")

class Child(Father, Mother):
    def show_skills(self):
        print("Child's skills:")
        Father.skills(self)
        Mother.skills(self)

# Usage
child_instance = Child()
child_instance.show_skills()  # Accessing methods from both parents

Child's skills:
Father's skills: Driving, Carpentry
Mother's skills: Cooking, Sewing


**3. Multilevel Inheritance**

In multilevel inheritance, a class is derived from another derived class, creating a chain of inheritance.


In [ ]:
#Example:

class Grandparent:
    def display(self):
        print("This is the Grandparent class.")

class Parent(Grandparent):
    def display(self):
        print("This is the Parent class.")

class Child(Parent):
    def display(self):
        print("This is the Child class.")

# Usage
child_instance = Child()
child_instance.display()  # Displays: This is the Child class.

This is the Child class.


**4. Hierarchical Inheritance**

Hierarchical inheritance occurs when multiple subclasses inherit from a single base class. This structure allows shared functionality among different derived classes.

In [ ]:
#Example:

class Animal:
    def speak(self):
        print("Animal speaks")

class Dog(Animal):
    def bark(self):
        print("Dog barks")

class Cat(Animal):
    def meow(self):
        print("Cat meows")

# Usage
dog_instance = Dog()
cat_instance = Cat()
dog_instance.speak()  # Method from Animal
cat_instance.speak()  # Method from Animal

Animal speaks
Animal speaks


**5. Hybrid Inheritance**

Hybrid inheritance is a combination of two or more types of inheritance, mixing elements from single, multiple, multilevel, and hierarchical inheritance.



In [ ]:
#Example:

class Base:
    def base_method(self):
        print("Base method")

class Derived1(Base):
    def derived1_method(self):
        print("Derived1 method")

class Derived2(Base):
    def derived2_method(self):
        print("Derived2 method")

class MultiDerived(Derived1, Derived2):
    def multi_method(self):
        print("MultiDerived method")

# Usage
obj = MultiDerived()
obj.base_method()      # Inherited from Base
obj.derived1_method()  # Inherited from Derived1
obj.derived2_method()  # Inherited from Derived2
obj.multi_method()     # MultiDerived's own method

Base method
Derived1 method
Derived2 method
MultiDerived method


**Summary:**

**Single Inheritance: **
Involves one base class and one derived class.

**Multiple Inheritance:**
A single derived class inherits from multiple base classes.


**Multilevel Inheritance:**
Involves a chain of classes where one class inherits from another derived class.

**Hierarchical Inheritance:**
Multiple derived classes stem from a single base class.

**Hybrid Inheritance:**
Combines various inheritance types to create a flexible class structure.

These inheritance types facilitate better organization of code, enhancing readability and maintainability.

#Question 7. What is the Method Resolution Order (MRO) in Python? How can you retrieve it programmatically?

#Answer:

The Method Resolution Order (MRO) in Python is a critical mechanism that dictates the order in which classes are searched when a method is called, particularly in situations involving multiple inheritance. The MRO ensures a consistent and predictable approach to method resolution, helping to avoid ambiguity when classes inherit from more than one parent class.


Python employs the C3 linearization algorithm to determine the MRO, which organizes the order of class searches.

How to Programmatically Retrieve the MRO
You can access the MRO of a class using either the __mro__ attribute or the mro() method. Here’s how to use both approaches:

In [ ]:
#1. Accessing the __mro__ Attribute:

class A:
    pass

class B(A):
    pass

class C(A):
    pass

class D(B, C):
    pass

print(D.__mro__)


#This will display the MRO for the class D, showing the sequence in which Python looks for methods.

(<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>)


#2. Using the mro() Method:

print(D.mro())


Both approaches will yield the same output, reflecting the order of classes in the MRO.


**Sample Output**
For the example above, the output will appear as follows:

(<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>)


This result indicates that when searching for a method in an instance of D, Python will first check D, followed by B, then C, and finally A, before reaching the base object class.


#Question 8. Create an abstract base class `Shape` with an abstract method `area()`. Then create two subclasses`Circle` and `Rectangle` that implement the `area()` method.

#Answer:

Below is an example of how to create an abstract base class Shape with an abstract method area(), and then implement two subclasses Circle and Rectangle that provide their own implementations of the area() method.


In [ ]:
from abc import ABC, abstractmethod
import math

# Abstract Base Class
class Shape(ABC):
    @abstractmethod
    def area(self):
        #Calculate the area of the shape
        pass

In [ ]:
# Circle Subclass
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
         #Return the area of the circle
        return math.pi * self.radius ** 2

In [ ]:
# Rectangle Subclass
class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        #Return the area of the rectangle
        return self.width * self.height
# Example usage
if __name__ == "__main__":
    circle = Circle(5)
    rectangle = Rectangle(4, 6)

    print(f"Circle Area: {circle.area():.2f}")
    print(f"Rectangle Area: {rectangle.area():.2f}")

Circle Area: 78.54
Rectangle Area: 24.00


**Explanation:**

**1.Shape Abstract Base Class:**

Inherits from ABC and defines an abstract method area() that must be implemented by subclasses.

**2.Circle Class:**

Represents a circle, initialized with a radius, and implements the area() method to compute the area using the formula
𝜋𝑟2


**3.Rectangle Class:**

Represents a rectangle, initialized with width and height, and implements the area() method to calculate the area as
width×height


**4.Main Block:**

Demonstrates how to create instances of Circle and Rectangle and prints their respective areas.



#Question 9. Demonstrate polymorphism by creating a function that can work with different shape objects to calculate and print their areas.

#Answer:

Through a common interface, polymorphism enables various objects to be considered as instances of the same class. We can construct a basic class named Shape with an area calculation function in the context of shapes. Then, by deriving from the Shape class and implementing their own area calculation, we may create distinct shape classes (such as Circle, Rectangle, and Triangle).

In [ ]:
import math

# Base class for shapes
class Shape:
    def area(self):
        raise NotImplementedError("This method should be overridden by subclasses.")

# Class representing a circle
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return math.pi * (self.radius ** 2)

# Class representing a rectangle
class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

# Class representing a triangle
class Triangle(Shape):
    def __init__(self, base, height):
        self.base = base
        self.height = height

    def area(self):
        return 0.5 * self.base * self.height

# Function to calculate and display the area of a shape
def display_area(shape):
    print(f"The area of the {shape.__class__.__name__} is: {shape.area()}")

# Creating instances of various shapes
shapes = [
    Circle(radius=4),
    Rectangle(width=4, height=5),
    Triangle(base=5, height=6)
]

# Looping through shapes to print their areas
for shape in shapes:
    display_area(shape)


The area of the Circle is: 50.26548245743669
The area of the Rectangle is: 20
The area of the Triangle is: 15.0


**Breakdown:**


**Base Class (Shape):**This serves as a template for all shape types, requiring derived classes to implement the area method.

**Derived Classes:**Each shape class (Circle, Rectangle, Triangle) implements its specific method to calculate the area based on its geometric properties.

**Function (display_area):**
This function receives a shape object and prints its area. It can accept any object that is derived from the Shape class.


**Output:**
When executed, the code will output:

The area of the Circle is: 50.26548245743669
The area of the Rectangle is: 20
The area of the Triangle is: 15.0


This example clearly illustrates polymorphism by allowing the display_area function to handle different shapes without needing to know their exact types. Each shape object correctly computes its area using the appropriate method.


#Question 10.  Implement encapsulation in a `BankAccount` class with private attributes for `balance` and `account_number`. Include methods for deposit, withdrawal, and balance inquiry

#Answer:

An easy Python implementation of the BankAccount class that shows encapsulation by utilising private attributes for account_number and balance. Techniques for making deposits, withdrawals, and balance checks are covered throughout the course.

In [ ]:
class BankAccount:
    def __init__(self, account_number, initial_balance=0):
        self.__account_number = account_number  # Private attribute for account number
        self.__balance = initial_balance  # Private attribute for balance

    def deposit(self, amount):
        """Add a specified amount to the account balance."""
        if amount > 0:
            self.__balance += amount
            print(f"Successfully deposited: ${amount:.2f}. Updated balance: ${self.__balance:.2f}.")
        else:
            print("The deposit amount must be greater than zero.")

    def withdraw(self, amount):
        """Withdraw a specified amount from the account if sufficient funds are available."""
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"Successfully withdrew: ${amount:.2f}. Updated balance: ${self.__balance:.2f}.")
        elif amount > self.__balance:
            print("Withdrawal amount exceeds current balance.")
        else:
            print("The withdrawal amount must be greater than zero.")

    def check_balance(self):
        """Retrieve the current balance of the account."""
        return self.__balance

    def get_account_number(self):
        """Get the account number associated with the account."""
        return self.__account_number

# Example usage:
if __name__ == "__main__":
    account = BankAccount("123456", 2000)
    account.deposit(400)
    account.withdraw(300)
    print(f"Account Number: {account.get_account_number()}")
    print(f"Current Balance: ${account.check_balance():.2f}")


Successfully deposited: $400.00. Updated balance: $2400.00.
Successfully withdrew: $300.00. Updated balance: $2100.00.
Account Number: 123456
Current Balance: $2100.00



#Question 11. Write a class that overrides the `__str__` and `__add__` magic methods. What will these methods allow you to do?

#Answer:

**Custom Class Example:**
Let’s create a class named Vector that represents a mathematical vector. We will override the __str__ method for a more meaningful string representation and the __add__ method to specify how two vector instances can be added together.

In [ ]:
class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        # This method is invoked by print() or str() for string representation
        return f"Vector({self.x}, {self.y})"

    def __add__(self, other):
        # This method defines how the + operator works for Vector instances
        if isinstance(other, Vector):
            return Vector(self.x + other.x, self.y + other.y)
        return NotImplemented  # Return NotImplemented if 'other' is not a Vector

# Example usage
v1 = Vector(4, 5)
v2 = Vector(6, 7)

print(v1)  # Output: Vector(4, 5)
print(v2)  # Output: Vector(6, 7)

v3 = v1 + v2
print(v3)  # Output: Vector(10, 12)


Vector(4, 5)
Vector(6, 7)
Vector(10, 12)


**Explanation of the Methods**

**1. __str__ Method:**

This method is automatically called when an object of the class is printed or when the str() function is applied to it.
It allows you to define how the object is represented as a string, making the output more informative and easier to read.
In our example, invoking print(v1) yields Vector(2, 3) instead of a default representation that is less descriptive, like <__main__.Vector object at 0x...>.



**2. __add__ Method:**

This method specifies how the + operator behaves when applied to instances of the class.
In this case, adding two Vector instances combines their respective x and y coordinates, resulting in a new Vector instance.
If you try to add a Vector to an incompatible type, the method returns NotImplemented, indicating that the operation is not supported.



**Benefits of Overriding These Methods:**

**Improved Readability:**
Customizing the __str__ method enhances the clarity of your objects when printed.

**Defined Operations:**
By overriding __add__, you can specify how instances interact with arithmetic operations, making them more intuitive to use.

**Seamless Integration:**
These custom implementations align the class with Python's built-in features, enhancing code usability and maintainability.

This pattern can be applied to other magic methods as well, allowing for further customization of class behavior.

#Question 12. Create a decorator that measures and prints the execution time of a function.

#Answer:

With Python's time module, you can write a decorator that calculates and outputs a function's execution time. Here's how to do it simply:

In [ ]:
import time

def measure_time(func):
    """Decorator to track and display the execution duration of a function."""
    def wrapper(*args, **kwargs):
        start = time.time()  # Record the start time
        result = func(*args, **kwargs)  # Execute the original function
        end = time.time()  # Record the end time
        duration = end - start  # Calculate the total time taken
        print(f"Execution time for '{func.__name__}': {duration:.6f} seconds")
        return result  # Return the result of the function call
    return wrapper

# Sample function to demonstrate the decorator
@measure_time
def sample_function(n):
    """Function that performs a summation up to n."""
    total = sum(range(n))
    return total

# Example of calling the sample function
sample_function(1000000)


Execution time for 'sample_function': 0.018909 seconds


499999500000

**Breakdown:**

**Decorator (measure_time)**:
This function takes another function as an argument and creates a nested wrapper function. The wrapper manages the execution and timing.

**Timing Logic**:
It captures the current time before and after the function call using time.time(), then calculates the elapsed time.

**Output**:
After the function completes, it prints the execution time, formatted to six decimal places.

**How to Use**:
You can apply the @measure_time decorator to any function whose execution time you wish to measure. In the provided example, sample_function will display how long it takes to execute when called.



#Question 13. Explain the concept of the Diamond Problem in multiple inheritance. How does Python resolve it?

#Answer:

The Diamond Problem is a challenge encountered in object-oriented programming, particularly with multiple inheritance. It occurs when a class inherits from two other classes that both derive from a shared superclass, creating a diamond-shaped inheritance structure. Here’s a diagram to illustrate this concept:

        A
       / \
      B   C
       \ /
        D

**In this diagram:**

Class A is the top-level superclass.
Classes B and C both inherit from A.
Class D inherits from both B and C.

**The Challenge**
The problem arises when a method is called on class D. If both B and C override a method from A, it creates ambiguity about which method to execute. This can result in several issues:

**Ambiguity:**
It is unclear whether to invoke the method from B or C.

**Code Duplication:**
If both B and C provide the same method implementation, this can lead to redundancy in the codebase.


**Python's Approach: Method Resolution Order (MRO)**

Python addresses the Diamond Problem through a mechanism known as Method Resolution Order (MRO), which specifies the order in which base classes are searched for methods. The MRO is determined using an algorithm called C3 linearization.

**Here’s how Python manages this situation:**

**1. Inheritance Sequence:**
For single inheritance, Python follows a straightforward order.

**2. MRO Determination:** When dealing with multiple inheritance, Python calculates the MRO for a class using the C3 linearization algorithm. You can access the MRO of a class using the __mro__ attribute or the mro() method.

In [ ]:
#Example : Here’s an example to demonstrate this:

class A:
    def greet(self):
        print("Hello from A")

class B(A):
    def greet(self):
        print("Hello from B")

class C(A):
    def greet(self):
        print("Hello from C")

class D(B, C):
    pass

d = D()
d.greet()  # Output: Hello from B
print(D.mro())

# Output: [<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>]

Hello from B
[<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>]


**Explanation of the Example**

In this example, when d.greet() is called, Python first searches in class D, then in B, where it finds the method and executes it. If the method were not found in B, Python would then check C, and finally A. The MRO for D illustrates the order in which classes are searched for method definitions, effectively resolving the ambiguity of the Diamond Problem.


To summarize, the Diamond Problem arises from the confusion surrounding method resolution in multiple inheritance scenarios. Python effectively resolves this issue using a well-defined method resolution order (MRO), ensuring that method lookups occur in a predictable and unambiguous manner.



#Question 14. Write a class method that keeps track of the number of instances created from a class.

#Answer:

Example in Python:



In [ ]:
class Counter:
    # Class-level variable to track the total number of instances
    total_instances = 0

    def __init__(self):
        # Increment the total instances count whenever a new object is instantiated
        Counter.total_instances += 1

    @classmethod
    def get_total_instances(cls):
        """Class method to retrieve the current count of instances."""
        return cls.total_instances


# Example of how to use the class
if __name__ == "__main__":
    instance1 = Counter()
    instance2 = Counter()
    instance3 = Counter()

    print("Total instances created:", Counter.get_total_instances())


Total instances created: 3


**Explanation:**

**Class Variable:**
total_instances is defined as a class variable, making it shared among all instances of the class.

**Initializer (__init__ method):**
This method is called when a new object is created, incrementing the total_instances count.

**Class Method (get_total_instances):**
This method allows you to access the count of instances from the class itself.

When executed, this code will display the total number of Counter instances created.


#Question 15. Implement a static method in a class that checks if a given year is a leap year.

#Answer:

In [ ]:
#Example:

class YearUtility:
    @staticmethod
    def check_leap_year(year):
        """
        Determine if a specified year is a leap year.

        The rules for a leap year are as follows:
        - A year is a leap year if it is divisible by 4.
        - It is not a leap year if it is divisible by 100, unless it is also divisible by 400.

        Parameters:
        year (int): The year to evaluate.

        Returns:
        bool: Returns True if the year is a leap year, otherwise False.
        """
        if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
            return True
        return False

# Example of how to use the class:
year = 2024
if YearUtility.check_leap_year(year):
    print(f"{year} is a leap year.")
else:
    print(f"{year} is not a leap year.")


2024 is a leap year.


In [ ]:
=============================THE-END==========